src/components/Api.js

```diff
@@ -0,0 +1,114 @@
+export default class API {
+  constructor({ baseUrl, headers }) {
+    this._baseUrl = baseUrl;
+    this._headers = headers;
+  }
+
+  _handleResponse(res) {
+    if (res.ok) {
+      return res.json();
+    } else {
+      return Promise.reject(`Error ${res.status}`);
+    }
+  }
+
+  _request(url, options) {
+    return fetch(url, options).then(this._handleResponse);
+  }
```

Comment:
**COULD BE IMPROVED**

You can make a universal method `_request` not to duplicate the `ok` checking, `headers` and `baseUrl`


```js
 _request(endpoint, options = {}) {
    const finalOptions = {
      // add the headers
      headers: this._headers,
      // spread the other options if provided
      ...options,
    };
      // add the endpoint to the base url
    const url = `${this._baseUrl}${endpoint}`;
    return fetch(url, finalOptions).then(this._checkResponse);
  }

  // usage

  getUserInfo() {
    // pass only the endpoint here
    return this._request("/users/me");
  }

  getInitialCards() {
    // pass only the endpoint here
    return this._request("/cards");
  }

  // and so on



```

Reply:
I was able to make this work for getUserUnfo and getInitialCards, the rest of the functions I wasn't sure how to go about adding the additional information, is that what you meant by spreading the other options provided? I would appreciate an example.

Reply:
Here is an example of `_request` usage

```js


  _request(url, options) {
    return fetch(url, options).then(this._handleResponse);
  }

  getUserInfo() {
    return this._request(`${this._baseUrl}/users/me`, {
      headers: this._headers,
    });
  }

  getInitialCards() {
    return this._request(`${this._baseUrl}/cards`, {
      headers: this._headers,
    });
  }

  updateProfileInfo(data) {
    return this._request(`${this._baseUrl}/users/me`, {
      method: "PATCH",
      headers: this._headers,
      body: JSON.stringify({
        name: data.title,
        about: data.description,
        avatar: data.avatar,
      }),
    });
  }
```

Reply:
> would appreciate an example.

An example of my variant of `_request`

```js

 _request(endpoint, options = {}) {
    const finalOptions = {
      // add the headers
      headers: this._headers,
      // spread the other options if provided
      ...options,
    };
      // add the endpoint to the base url
    const url = `${this._baseUrl}${endpoint}`;
    return fetch(url, finalOptions).then(this._checkResponse);
  }

  // usage

  getUserInfo() {
    // pass only the endpoint here
    return this._request("/users/me");
  }

  getInitialCards() {
    // pass only the endpoint here
    return this._request("/cards");
  }

  // this is your method with `_request`
  updateProfileInfo(data) {
    return this._request(`/users/me`, {
      method: "PATCH",
      body: JSON.stringify({
        name: data.title,
        about: data.description,
        avatar: data.avatar,
      }),
    });
  }



```